mobile_verifier/src/boosting_oracles/mod.rs

```diff
@@ -1,184 +1,600 @@
+        // If there is an unprocessed data set, download it (if we need to) and process it.
+
+        let path = self.get_data_set_path(latest_unprocessed_data_set.time_to_use);
+
+        // Download the file if it hasn't been downloaded already:
+        if !latest_unprocessed_data_set.status.is_downloaded() {
+            download_data_set(&self.store, &latest_unprocessed_data_set.filename, &path).await?;
+            db::set_data_set_status(
+                &self.pool,
+                &latest_unprocessed_data_set.filename,
+                DataSetStatus::Downloaded,
+            )
+            .await?;
+            tracing::info!(
+                data_set = latest_unprocessed_data_set.filename,
+                "Data set download complete"
+            );
+        }
+
+        // Now that we've downloaded the file, load it into the data set
+        data_set.update(Path::new(&path), latest_unprocessed_data_set.time_to_use)?;
+
+        // Release the lock as it is no longer needed
+        drop(data_set);
+
+        // Update the hexes
+        let boosting_reports = set_oracle_boosting_assignments(
+            UnassignedHex::fetch_all(&self.pool),
+            &self.data_sets,
+            &self.pool,
+        )
+        .await?;
+
+        db::set_data_set_status(
+            &self.pool,
+            &latest_unprocessed_data_set.filename,
+            DataSetStatus::Processed,
+        )
+        .await?;
+
+        self.oracle_boosting_sink
+            .write_all(boosting_reports)
+            .await?;
+        self.oracle_boosting_sink.commit().await?;
+        tracing::info!(
+            data_set = latest_unprocessed_data_set.filename,
+            "Data set processing complete"
         );
 
-        let cover = Landtype::try_from(vals[0])?;
-        Ok(cover.into())
+        Ok(())
+    }
+
+    pub async fn run(mut self) -> anyhow::Result<()> {
+        // Get the first data set:
+        if let Some(time_to_use) =
+            db::fetch_time_of_latest_processed_data_set(&self.pool, T::TYPE).await?
+        {
+            let data_set_path = self.get_data_set_path(time_to_use);
+            tracing::info!(
+                "Found initial {} data set: {}",
+                T::TYPE.to_prefix(),
+                data_set_path.to_string_lossy()
+            );
+            self.data_set
+                .lock()
+                .await
+                .update(&data_set_path, time_to_use)?;
+        }
+
+        let poll_duration = Duration::minutes(5);
+
+        loop {
+            self.check_for_available_data_sets().await?;
+            self.process_data_sets().await?;
+            tokio::time::sleep(poll_duration.to_std()?).await;
+        }
+    }
+}
+
+async fn download_data_set(
+    store: &FileStore,
+    in_file_name: &str,
+    out_path: &Path,
+) -> anyhow::Result<()> {
+    tracing::info!("Downloading new data set: {}", out_path.to_string_lossy());
+    // TODO: abstract this out to a function
```

Comment:
This todo still relevant ? looks not as download data set does what it says on the tin ?